### PR TITLE
Eagerly fail stream on failed future in Source.future

### PR DESCRIPTION
**Pull request by leviramsey**
The Scaladoc for `Source.future` says:

> The stream fails if the Future is completed with a failure.

However, in the current implementation, if there is never demand from downstream, the stream does not fail.

This is a behavior change: for instance in the current implementation:

```scala
Source.never
  .concat(Source.future(Future.failed(new RuntimeException))
  .runWith(Sink.ignore)
```

never fails, but would with this change.  If this is a big problem, I guess we can introduce a new `Source.eagerlyFailingFuture` (or whatever) with this implementation and leave `Source.future` as is (with updated Scaladoc to specifically say that the stream will only fail once there is demand on the source).

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala [hunk 1]
```diff
@@ -432,27 +432,45 @@ import scala.util.control.NonFatal
     override def initialAttributes: Attributes = DefaultAttributes.futureSource
     override def createLogic(attr: Attributes) =
       new GraphStageLogic(shape) with OutHandler {
+        // Always register a callback which will fail the stage if the future fails...
+        val cb = getAsyncCallback[Try[T]](onFutureCompleted).invoke _
+        future.onComplete(cb)(ExecutionContexts.parasitic)
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala:437]: Extract the code in `onPull` to a dedicated method , and then we can keep the optimization if the future is already completed.

**Comment by He-Pin** [comment]
it's lazy in `unfoldAsync` too.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala [hunk 2]
```diff
@@ -500,4 +500,12 @@ class SourceSpec extends StreamSpec with DefaultTimeout {
       Await.result(result, 4.seconds) shouldBe Done
     }
   }
+
+  "Source.future" must {
+    "fail on failed future, even with no demand" in {
+      val result = Source.future(Future.failed(new RuntimeException("BOOM!"))).runWith(Sink.never)
+
+      (the[RuntimeException] thrownBy (Await.result(result, 1.second))).getMessage shouldBe "BOOM!"
+    }
```
Comment (He-Pin) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala:509]: add to FlowFromFutureSpec?

**Comment by GerretS** [comment]
Hello, I made the original report in gitter. 

I have one followup question.

@leviramsey says:
> However, in the current implementation, if there is never demand from downstream, the stream does not fail.

From what I can tell it's not actually about demand from downstream. See this example:

```scala
val decider: Supervision.Decider = { _ =>
  println("HELLO")
  Supervision.Stop
}

implicit val as: ActorSystem = ActorSystem()

val fut = Future.failed[Unit](new RuntimeException("test"))

val source = Source.future(fut).withAttributes(ActorAttributes.supervisionStrategy(decider))
val sourceFut = source.runWith(Sink.foreach(x => println(s"Sink $x")))

Await.ready(sourceFut, 1.minute)

println("Done")
```

Here, println("HELLO") in the devider is never triggered, even though I make sure there is a Sink.

If I change the definition of my source to instead do the future in a mapAsync, `HELLO` is printed.
```scala
val source = Source.single(()).mapAsync(1)(_ => fut).withAttributes(ActorAttributes.supervisionStrategy(decider))
```

Is this situation also handled by this fix?

**Comment by He-Pin** [comment]
Noop, I think you would like to use is:
```scala
def createFuture() = ???
Source.future(createFuture).recoverWith({case _ => Source.future(createFuture))
```
or by using `lazyFuture`?

**Comment by GerretS** [comment]
Then there's an interesting bit of miscommunication here. My original question was whether it was expected behaviour that there is no supervision on Source.future even though the documentation sentence `The stream fails if the Future is completed with a failure.` made me think there is, and then @leviramsey actually went and fixed a different but related issue.

**Comment by He-Pin** [comment]
I think this PR is still valid and good to go, for your origin issue, I think that's expected too.
Akka does not mention which operator implemented the supervision strategy , so that causes some superizing sometime.

**Comment by johanandren** [comment]
The stages that does adhere to a supervision strategy are (or if they aren't should be) explicitly marked with 
> * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
In the Java/Scaladoc, it is not expected that all stages adheres. 

I think this one should stay as is wrt the supervision strategy since the Future handed to it is not necessarily in its control/owned vy it, can be recovered in a few different way, and also that not all supervision strategies makes sense for it.

Also not sure we should change behavior wrt the eager fail @leviramsey I think I'd rather see we change the documentation, to not break existing stream code.

**Comment by leviramsey** [comment]
I guess the question is whether it's more likely that some existing code depends on the incorrect behavior or incorrectly assumes the correct behavior.

**Comment by johanandren** [comment]
My thinking: 
 * If you depend on a stream failing early that keeps backpressuring you'd notice because it is stuck and doesn't fail.
 * If we change it a stream that currently cancels after only backpressuring could start failing, while currently successfully completing.

**Comment by He-Pin** [comment]
I think the eagarly failing can alian with the current `FailedSource` behavior which immediately ends the stream.

**Comment by leviramsey** [comment]
Waiting for the mima failures :D

**Comment by johanandren** [comment]
I don't think it should be a toggle/opt in since it can just as well be composed with detach.

**Comment by johanandren** [comment]
No discussion here for a while, I think it will break code if we change the behaviour, so let's close it.

**Pull request closed without merging**